File: frontend/types/content.ts
/**
 * Content types aligned with Schema v3 structure
 * Supporting multi-source content (PubMed, YouTube, GitHub, Discourse, Wiki)
 */

export interface Author {
  name: string
  email?: string
  affiliation?: string
  orcid?: string
}

export interface ContentMetrics {
  view_count: number
  bookmark_count: number
  share_count: number
  citation_count: number
}

export interface Citations {
  cited_by_count: number
  references_count: number
  cited_by_ids: string[]
  reference_ids: string[]
}

export interface AIEnhancement {
  ai_enhanced: boolean
  ai_confidence?: number  // Schema v3: replaces ai_quality_score
  ai_summary?: string
  ai_tools?: string[]
  ai_categories?: string[]
}

// Base content interface with all source-specific fields
export interface ContentItem extends YouTubeFields, GitHubFields, DiscourseFields, WikiFields, PubMedFields {
  id: string
  title: string
  description?: string
  abstract?: string
  content?: string
  content_type: string
  
  // Source information
  source: 'pubmed' | 'youtube' | 'github' | 'discourse' | 'wiki'
  display_type?: string  // Computed field
  icon_type?: string     // Computed field
  content_category?: string  // Computed field
  
  // Authors and metadata
  authors: Author[]
  
  // Dates
  published_date?: string
  created_at?: string
  updated_at?: string
  last_activity?: string
  last_modified?: string
  
  // Categories and scoring - Schema v3
  categories: string[]
  final_score?: number  // Schema v3: replaces combined_score
  ml_score?: number
  
  // Metrics
  metrics: ContentMetrics
  
  // AI Enhancement - Schema v3
  ai_enhanced?: boolean
  ai_confidence?: number
  ai_summary?: string
  ai_tools?: string[]
  ai_categories?: string[]
  
  // URLs
  url?: string
  thumbnail_url?: string
  
  // Search highlighting
  highlight?: Record<string, string[]>
  
  // Citations
  citations?: Citations
}

// Source-specific fields are included in base ContentItem interface
// YouTube-specific fields
export interface YouTubeFields {
  video_id?: string
  duration?: number
  channel_name?: string
  thumbnail_url?: string
}

// GitHub-specific fields
export interface GitHubFields {
  repo_name?: string
  stars_count?: number
  forks_count?: number
  language?: string
  topics?: string[]
  owner?: string
  last_commit?: string
}

// Discourse-specific fields
export interface DiscourseFields {
  topic_id?: number
  reply_count?: number
  solved?: boolean
  category?: string
  tags?: string[]
}

// Wiki-specific fields
export interface WikiFields {
  doc_type?: string
  section_count?: number
  read_time?: number
}

// PubMed-specific fields
export interface PubMedFields {
  journal?: string
  institutions?: string[]
  mesh_terms?: string[]
  pmid?: string
  doi?: string
  year?: number
}

// Union type for all content types (simplified since ContentItem includes all fields)
export type AnyContentItem = ContentItem

// Sort options
export type SortOption = 'relevance' | 'date' | 'popularity' | 'score' | 'title'

// View modes
export type ViewMode = 'compact' | 'expanded' | 'grouped'
export type LayoutMode = 'list' | 'grid'

// Filter options
export interface ContentFilters {
  sources?: Array<'pubmed' | 'youtube' | 'github' | 'discourse' | 'wiki'>
  categories?: string[]
  dateRange?: {
    start?: string
    end?: string
  }
  scoreRange?: {
    min?: number
    max?: number
  }
  contentTypes?: string[]
  languages?: string[]
}

// Search result structure
export interface SearchResult {
  total: number
  items: ContentItem[]
  aggregations: Record<string, any>
  took_ms: number
}

// Error types
export interface ContentError {
  message: string
  code?: string
  details?: any
}

// Loading state
export interface LoadingState {
  loading: boolean
  error?: ContentError
  progress?: number
}

File: backend/app/api/graphql/types/content.py
"""
Content-related GraphQL types.

Includes Author, ContentMetrics, Content, and SearchResult types.
"""
import strawberry
from strawberry.scalars import JSON
from typing import Optional, List

from ..helpers import compute_display_fields
from ..services import search_service


@strawberry.type
class Author:
    name: str
    email: Optional[str] = None
    affiliation: Optional[str] = None
    orcid: Optional[str] = None  # Added for ES schema v3


@strawberry.type
class ContentMetrics:
    view_count: int = 0
    bookmark_count: int = 0
    share_count: int = 0
    citation_count: int = 0  # Added for ES schema v3


@strawberry.type
class Content:
    id: str
    title: str
    abstract: Optional[str]
    content_type: str

    # Multimodal fields
    source: Optional[str] = None  # pubmed, youtube, github, discourse, wiki
    display_type: Optional[str] = None  # Research Article, Video Content, Code Repository, etc.
    icon_type: Optional[str] = None  # document-text, play-circle, code, chat-bubble, book-open
    content_category: Optional[str] = None  # research, media, code, community, reference

    authors: List[Author]
    published_date: Optional[str]  # Changed to str to handle date-only format
    created_at: Optional[str] = None  # Changed to str for flexibility
    ml_score: Optional[float]
    ai_confidence: Optional[float] = None  # Consolidated from gpt_score
    final_score: Optional[float] = None  # Consolidated from combined_score
    categories: List[str]  # Consolidated categories field
    metrics: ContentMetrics
    url: Optional[str] = None
    journal: Optional[str] = None
    doi: Optional[str] = None
    keywords: List[str] = strawberry.field(default_factory=list)
    year: Optional[int] = None

    # Source-specific fields
    # YouTube
    video_id: Optional[str] = None
    duration: Optional[int] = None  # in seconds
    channel_name: Optional[str] = None
    thumbnail_url: Optional[str] = None
    transcript: Optional[str] = None

    # GitHub
    owner: Optional[str] = None
    repo_name: Optional[str] = None
    stars_count: Optional[int] = None
    watchers_count: Optional[int] = None
    forks_count: Optional[int] = None
    open_issues_count: Optional[int] = None
    contributors_count: Optional[int] = None
    contributors: List[str] = strawberry.field(default_factory=list)
    readme_content: Optional[str] = None
    language: Optional[str] = None
    license: Optional[str] = None
    topics: List[str] = strawberry.field(default_factory=list)  # Added for ES schema v3
    last_commit: Optional[str] = None  # Changed to str for flexibility

    # Discourse
    topic_id: Optional[str] = None
    reply_count: Optional[int] = None
    solved: Optional[bool] = None

    # Wiki
    doc_type: Optional[str] = None
    section_count: Optional[int] = None
    last_modified: Optional[str] = None  # Changed to str for flexibility

    # Citations (simplified structure for ES schema v3)
    pmid: Optional[str] = None
    citations: Optional[JSON] = None  # Simplified: {cited_by_count, references_count, cited_by_ids[], reference_ids[]}

    # PubMed specific fields (added for ES schema v3)
    journal: Optional[str] = None
    mesh_terms: List[str] = strawberry.field(default_factory=list)

    # AI Enhancement fields
    ai_enhanced: Optional[bool] = None
    ai_is_ohdsi: Optional[bool] = None
    ai_summary: Optional[str] = None
    ai_tools: List[str] = strawberry.field(default_factory=list)

    @strawberry.field
    async def related_content(self, limit: int = 5) -> List["Content"]:
        """Get related content"""
        related = await search_service.get_related(self.id, limit)
        related_items = []
        for item in related:
            # Process authors - only extract the fields we need
            authors = []
            if item.authors:
                for author_data in item.authors:
                    # Handle author data whether it's a dict or object
                    if hasattr(author_data, '__dict__'):
                        author_dict = author_data.__dict__
                    else:
                        author_dict = author_data

                    author = Author(
                        name=author_dict.get("name", ""),
                        email=author_dict.get("email"),
                        affiliation=author_dict.get("affiliation")
                    )
                    authors.append(author)

            # Compute display fields for related content
            computed_fields = compute_display_fields(getattr(item, 'source', ''), item.content_type)

            related_items.append(Content(
                id=item.id,
                title=item.title,
                abstract=item.abstract,
                content_type=item.content_type,

                # Multimodal fields (with computed fallbacks for ES schema v3)
                source=getattr(item, 'source', None),
                display_type=getattr(item, 'display_type', computed_fields['display_type']),
                icon_type=getattr(item, 'icon_type', computed_fields['icon_type']),
                content_category=getattr(item, 'content_category', computed_fields['content_category']),

                authors=authors,
                published_date=str(item.published_date) if item.published_date else None,
                ml_score=item.ml_score,
                categories=getattr(item, 'categories', []),
                # Handle metrics as either object or dict
                metrics=ContentMetrics(**item.metrics.__dict__) if hasattr(item.metrics, '__dict__') else ContentMetrics(**item.metrics) if item.metrics else ContentMetrics(),
                journal=getattr(item, 'journal', None),
                doi=getattr(item, 'doi', None),
                keywords=getattr(item, 'keywords', []),
                year=getattr(item, 'year', None)
            ))
        return related_items


@strawberry.type
class SearchResult:
    total: int
    items: List[Content]
    aggregations: strawberry.scalars.JSON
    took_ms: int
